src/interface/swarm.cpp

```diff
@@ -455,23 +432,35 @@ void Swarm::Defrag() {
 
   auto &intVector_ = std::get<getType<int>()>(Vectors_);
   auto &realVector_ = std::get<getType<Real>()>(Vectors_);
-  SwarmVariablePack<Real> vreal;
-  SwarmVariablePack<int> vint;
-  PackIndexMap rmap;
-  PackIndexMap imap;
-  vreal = PackAllVariables<Real>(rmap);
-  vint = PackAllVariables<int>(imap);
+  PackIndexMap real_imap;
+  PackIndexMap int_imap;
+  ParArrayND<int> rpack_indices_shapes;
+  ParArrayND<int> ipack_indices_shapes;
+  auto vreal = PackAllVariables<Real>(real_imap, rpack_indices_shapes);
+  auto vint = PackAllVariables<int>(int_imap, ipack_indices_shapes);
   int real_vars_size = realVector_.size();
   int int_vars_size = intVector_.size();
+  auto real_map = real_imap.Map();
+  auto int_map = int_imap.Map();
 
   pmb->par_for(
       "Swarm::DefragVariables", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
         if (from_to_indices(n) >= 0) {
           for (int i = 0; i < real_vars_size; i++) {
-            vreal(i, from_to_indices(n)) = vreal(i, n);
+            for (int j = 0; j < rpack_indices_shapes(1, i); j++) {
+              for (int k = 0; k < rpack_indices_shapes(2, i); k++) {
+                vreal(rpack_indices_shapes(0, i), k, j, from_to_indices(n)) =
+                    vreal(rpack_indices_shapes(0, i), k, j, n);
+              }
+            }
           }
           for (int i = 0; i < int_vars_size; i++) {
-            vint(i, from_to_indices(n)) = vint(i, n);
+            for (int j = 0; j < ipack_indices_shapes(1, i); j++) {
+              for (int k = 0; k < ipack_indices_shapes(2, i); k++) {
+                vint(ipack_indices_shapes(0, i), k, j, from_to_indices(n)) =
+                    vint(ipack_indices_shapes(0, i), k, j, n);
+              }
+            }
```

Comment:
I had troubles mentally parsing these lines here and then realized that this is/was likely related to me not being familiar with the packing of SwarmVariables.
If I didn't miss anything this is not documented yet, isn't it?
It'd be great to add some documentation both inline as well as in the `particles.md` file.

Reply:
Good point, thanks. `PackAllVariables` actually has a unique syntax really tailored just to filling buffers with all particle data for communication, so I've added inline comments describing the layout but also made it a private function. The `PackVariables` function is a traditional pack of `ParArrayND`s so I'll add some documentation to `particles.md` at least to say that it exists (It's not used in any examples right now).

Reply:
Just leaving a comment after I now realized why I got confused here. In my mind `k`, `j`, and `i` are kind of closely tied to "mesh" variables, so I got thrown off.
If more people have that "mental map" it may be worth discussing some notation (this also, for example, applies to the higher dimensional variable names used for mesh variables).

Reply:
Yes I agree, looking at this again `i, j, k` are unfortunate choices. I switched this to something hopefully clearer, `i -> vidx`, `j -> l`, `k -> m` (and switched the `l` and `m` loop ordering to match the index ordering in the arrays)